packages/astro/src/runtime/server/transition.ts

```diff
@@ -22,78 +21,80 @@ export function createTransitionScope(result: SSRResult, hash: string) {
 	return `astro-${hash}-${num}`;
 }
 
+// Ensure animationName is a valid CSS identifier
+function toValidIdent(name: string): string {
+	return name.replace(/[^a-zA-Z0-9\-\_]/g, '_').replace(/^\_+|\_+$/g, '')
+}
+
+type Entries<T extends Record<string, any>> = Iterable<[keyof T, T[keyof T]]>
+
+const getAnimations = (name: TransitionAnimationValue) => {
+	if (name === 'fade') return fade();
+	if (name === 'slide') return slide();
+	if (typeof name === 'object') return name;
+}
+
 export function renderTransition(
 	result: SSRResult,
 	hash: string,
 	animationName: TransitionAnimationValue | undefined,
 	transitionName: string
 ) {
-	let animations: TransitionDirectionalAnimations | null = null;
-	switch (animationName) {
-		case 'fade': {
-			animations = fade();
-			break;
-		}
-		case 'slide': {
-			animations = slide();
-			break;
-		}
-		default: {
-			if (typeof animationName === 'object') {
-				animations = animationName;
+	// Default to `fade` (similar to `initial`, but snappier)
+	if (!animationName) animationName = 'fade';
+	const scope = createTransitionScope(result, hash);
+	const name = transitionName ? toValidIdent(transitionName) : scope;
+	const sheet = new ViewTransitionStyleSheet(scope, name);
+	
+	const animations = getAnimations(animationName);
+	if (animations) {
+		for (const [direction, images] of Object.entries(animations) as Entries<typeof animations>) {
+			for (const [image, rules] of Object.entries(images) as Entries<typeof animations[typeof direction]>) {
+				sheet.addAnimationPair(direction, image, rules);
 			}
 		}
+	} else if (animationName === 'none') {
+		sheet.addAnimationRaw('old', 'animation: none; opacity: 0; mix-blend-mode: normal;')
+		sheet.addAnimationRaw('new', 'animation: none; mix-blend-mode: normal;')
 	}
 
-	const scope = createTransitionScope(result, hash);
+	result._metadata.extraHead.push(markHTMLString(`<style>${sheet.toString()}</style>`));
+	return scope;
+}
 
-	// Default transition name is the scope of the element, ie HASH-1
-	if (!transitionName) {
-		transitionName = scope;
-	}
+class ViewTransitionStyleSheet {
```

Comment:
Cleaned the style generation up significantly with a builer class—I highly recommend reviewing this file without the diff!